src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HttpConnectionPool.cs

```diff
@@ -648,6 +655,132 @@ public byte[] Http2AltSvcOriginUri
+
+                    if (IsAlive(http2Connection) && http2Connection.AcquireStreamSlot())
+                    {
+                        return http2Connection;
+                    }
+                }
+
+                // No connection with a free stream slot found.
+
+                if (localConnections.Length < _poolManager.Settings._maxHttp2ConnectionsPerServer)
+                {
+                    // Procceed to a new connection construction.
+                    return null;
+                }
+
+                // Connection maximum is reached. Wait for a free slot on any existing connection.
+
+                SemaphoreSlim? localfreeConnections;
+                lock (SyncObj)
+                {
+                    _waitingRequestCount++;
+                    if (_freeStreamSlots == null)
+                    {
+                        _freeStreamSlots = new SemaphoreSlim(0);
+                    }
+                    localfreeConnections = _freeStreamSlots;
+                }
+
+                await localfreeConnections.WaitAsync().ConfigureAwait(false);
+
+                lock (SyncObj)
+                {
+                    _waitingRequestCount--;
+                    if (_waitingRequestCount == 0)
+                    {
+                        // No more waiting requests. Stop tracking available stream slots.
+                        _freeStreamSlots = null;
+                    }
+                }
+            }
+        }
+
+        private bool IsAlive(Http2Connection http2Connection)
+        {
+            TimeSpan pooledConnectionLifetime = _poolManager.Settings._pooledConnectionLifetime;
+            if (http2Connection.LifetimeExpired(Environment.TickCount64, pooledConnectionLifetime))
+            {
+                // Connection expired.
+                http2Connection.Dispose();
+                InvalidateHttp2Connection(http2Connection);
+                return false;
+            }
+            return true;
+        }
+
+        private void AddHttp2Connection(Http2Connection newConnection)
+        {
+            lock (SyncObj)
+            {
+                Http2Connection[]? localHttp2Connections = _http2Connections;
+                int newCollectionSize = localHttp2Connections == null ? 1 : localHttp2Connections.Length + 1;
+                Http2Connection[] newHttp2Connections = new Http2Connection[newCollectionSize];
+                newHttp2Connections[0] = newConnection;
+
+                if (localHttp2Connections != null)
+                {
+                    Array.Copy(localHttp2Connections, 0, newHttp2Connections, 1, localHttp2Connections.Length);
+                }
+
+                _http2Connections = newHttp2Connections;
+            }
+        }
+
+        public void StreamSlotAvailable()
+        {
+            if (EnableMultipleHttp2Connections)
+            {
+                SemaphoreSlim? localFreeConnections = Volatile.Read(ref _freeStreamSlots);
```

Comment:
What prevents race conditions where this reads null and thus doesn't call Release, and immediately after that a requester  creates the semaphore and waits on it?

Reply:
Fixed.